external/llvm-project/mlir/lib/Transforms/Utils/OutlinerUtils.cpp

```diff
@@ -173,337 +209,380 @@ static bool opsMatch(Operation *lhs, Operation *rhs, OutliningCandidate &one,
+      if (std::get<0>(params) != std::get<1>(params)) {
+        return false;
+      }
+    }
+    // get Result Types
+
+    for (auto opPair : llvm::zip(ops, that.ops)) {
+      if (!opsMatch(std::get<0>(opPair), std::get<1>(opPair), that)) {
+        return false;
+      }
+    }
+    return true;
   }
 
-  if (one.params.size() != two.params.size()) {
-    return false;
+  return false;
+}
+
+////////////////////////////////////////////////////////////////////////////////////
+////   OutlineBuilder
+////   1) Collect all ops that can be fused
+////   2) Look for match in Candidate Set
+////   3) If no match, build outline func
+////   4) Replace ops with call to outline func
+////////////////////////////////////////////////////////////////////////////////////
+
+class OutlineBuilder {
+
+  // Debug messaging for traversal
+  inline void debug(const char *tag, Operation *op) const {
+    LLVM_DEBUG(llvm::dbgs() << tag << ": " << op << "\n");
   }
-  for (auto params : llvm::zip(one.params, two.params)) {
-    if (std::get<0>(params) != std::get<1>(params)) {
-      return false;
+
+  inline void debug(const char *tag, Value v) const {
+    LLVM_DEBUG(llvm::dbgs() << tag << ": "; v.print(
+        llvm::dbgs(), OpPrintingFlags().elideLargeElementsAttrs());
+               llvm::dbgs() << "\n");
+  }
+
+  // Add Value/Operation methods
+  void addInput(Value v) {
+    debug("INPUT", v);
+    _inputs.insert(v);
+  }
+
+  void addResult(Value v) {
+    debug("RESULT", v);
+    _results.insert(v);
+  }
+
+  void addPredecessor(Operation *op) {
+    if (!_leadingOps.contains(op)) {
+      assert(!_trailingOps.contains(op));
+      debug("PRED", op);
+      _leadingOps.insert(op);
+      _locs.push_back(op->getLoc());
     }
   }
 
-  for (auto ops : llvm::zip(one.leadingOps, two.leadingOps)) {
-    if (!opsMatch(std::get<0>(ops), std::get<1>(ops), one, two)) {
-      return false;
+  void addSuccessor(Operation *op) {
+    if (!_trailingOps.contains(op)) {
+      debug("SUCC", op);
+      _trailingOps.insert(op);
+      _locs.push_back(op->getLoc());
     }
   }
-  if (!opsMatch(one.anchorOp, two.anchorOp, one, two)) {
-    return false;
+
+  bool contains(Operation *op) const {
+    return _trailingOps.contains(op) || _leadingOps.contains(op);
   }
-  for (auto ops : llvm::zip(one.trailingOps, two.trailingOps)) {
-    if (!opsMatch(std::get<0>(ops), std::get<1>(ops), one, two)) {
+
+  // Recurse back full chain looking for anchorOp
+  // - uses encountered set to reduce time
+  bool inChainToAnchor(Operation *inOp,
+                       DenseSet<Operation *> &encountered) const {
+    if (inOp == nullptr)
+      return false;
+    else if (inOp == _anchorOp)
+      return true;
+    else if (encountered.contains(inOp))
       return false;
+    encountered.insert(inOp);
+    for (auto opr : inOp->getOperands()) {
+      if (Operation *oprOp = opr.getDefiningOp()) {
+        debug(">>>", oprOp);
```

Comment:
I think there's some magic incantations for showing recursion on a tree in the debug stream - it might be in the pattern rewriters somewhere. Or maybe I'm way off.

Reply:
I'm having trouble understanding what this is for.  Wouldn't any op between the current and the anchor be included in the trailing-ops set?  When traversing forward from anchor, an op that doesn't qualify as trailing should stop the traversal.

Depending on circumstances, this could also be asking a post-dominance question.

Reply:
This fixes the divergence problem. For example:

                         anchor
                               |
                           mul
                          /      \
                          |     add
                          |        |
                          |      mul
                          |         |
                          \       /
                           add
                             | 
                            ...

Previous impl would stop at the mul and the subsequent add and have 2 outputs. This code will allow an op to be collected if all it's inputs are external or immediately available in the collection.

Reply:
Hmm, got formatted.. does it make sense?